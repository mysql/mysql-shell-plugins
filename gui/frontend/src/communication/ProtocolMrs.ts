/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

export enum ShellAPIMrs {
    //  Begin auto generated API names
    MrsAddService = "mrs.add.service",
    MrsGetService = "mrs.get.service",
    MrsListServices = "mrs.list.services",
    MrsEnableService = "mrs.enable.service",
    MrsDisableService = "mrs.disable.service",
    MrsDeleteService = "mrs.delete.service",
    MrsSetServiceDefault = "mrs.set.service.default",
    MrsSetServiceContextPath = "mrs.set.service.context_path",
    MrsSetServiceProtocol = "mrs.set.service.protocol",
    MrsSetServiceComments = "mrs.set.service.comments",
    MrsSetServiceOptions = "mrs.set.service.options",
    MrsUpdateService = "mrs.update.service",
    MrsGetServiceRequestPathAvailability = "mrs.get.service_request_path_availability",
    MrsAddSchema = "mrs.add.schema",
    MrsGetSchema = "mrs.get.schema",
    MrsListSchemas = "mrs.list.schemas",
    MrsEnableSchema = "mrs.enable.schema",
    MrsDisableSchema = "mrs.disable.schema",
    MrsDeleteSchema = "mrs.delete.schema",
    MrsSetSchemaName = "mrs.set.schema.name",
    MrsSetSchemaRequestPath = "mrs.set.schema.request_path",
    MrsSetSchemaRequiresAuth = "mrs.set.schema.requires_auth",
    MrsSetSchemaItemsPerPage = "mrs.set.schema.items_per_page",
    MrsSetSchemaComments = "mrs.set.schema.comments",
    MrsUpdateSchema = "mrs.update.schema",
    MrsAddContentSet = "mrs.add.content_set",
    MrsListContentSets = "mrs.list.content_sets",
    MrsGetContentSet = "mrs.get.content_set",
    MrsEnableContentSet = "mrs.enable.content_set",
    MrsDisableContentSet = "mrs.disable.content_set",
    MrsDeleteContentSet = "mrs.delete.content_set",
    MrsAddDbObject = "mrs.add.db_object",
    MrsGetDbObject = "mrs.get.db_object",
    MrsGetDbObjectRowOwnershipFields = "mrs.get.db_object_row_ownership_fields",
    MrsGetDbObjectFields = "mrs.get.db_object_fields",
    MrsListDbObjects = "mrs.list.db_objects",
    MrsGetDbObjectParameters = "mrs.get.db_object_parameters",
    MrsSetDbObjectRequestPath = "mrs.set.db_object.request_path",
    MrsSetDbObjectCrudOperations = "mrs.set.db_object.crud_operations",
    MrsEnableDbObject = "mrs.enable.db_object",
    MrsDisableDbObject = "mrs.disable.db_object",
    MrsDeleteDbObject = "mrs.delete.db_object",
    MrsUpdateDbObject = "mrs.update.db_object",
    MrsListContentFiles = "mrs.list.content_files",
    MrsGetAuthenticationVendors = "mrs.get.authentication_vendors",
    MrsAddAuthenticationApp = "mrs.add.authentication_app",
    MrsListAuthenticationApps = "mrs.list.authentication_apps",
    MrsInfo = "mrs.info",
    MrsVersion = "mrs.version",
    MrsLs = "mrs.ls",
    MrsConfigure = "mrs.configure",
    MrsStatus = "mrs.status",
    //  End auto generated API names
}
