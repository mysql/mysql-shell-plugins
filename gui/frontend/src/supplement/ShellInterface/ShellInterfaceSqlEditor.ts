/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { EventType, IDispatchEvent, ListenerEntry } from "../Dispatch";
import {
    ProtocolGui, ICommErrorEvent, ICommStartSessionEvent, ShellAPIGui, ICommSimpleResultEvent,
    ShellPromptResponseType, IPromptReplyBackend, MessageScheduler, IGenericResponse,
} from "../../communication";
import { webSession } from "../WebSession";
import { settings } from "../Settings/Settings";
import { ShellInterfaceDb, ShellInterfaceMds, ShellInterfaceMrs } from ".";

export class ShellInterfaceSqlEditor extends ShellInterfaceDb implements IPromptReplyBackend {

    public mds: ShellInterfaceMds = new ShellInterfaceMds();
    public mrs: ShellInterfaceMrs = new ShellInterfaceMrs();

    public get id(): string {
        return "sqlEditor";
    }

    /**
     * @returns a flag which indicates if a session was opened already.
     */
    public get hasSession(): boolean {
        return this.moduleSessionId !== undefined;
    }

    /**
     * Begins a new session for a specific SQL editor. This overrides the underlying DB session handling.
     *
     * @param id A value identifying the SQL editor.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public startSession(id: string): Promise<void> {
        return new Promise((resolve, reject) => {
            this.moduleSessionLookupId = this.id + "." + id;
            this.mrs.moduleSessionLookupId = this.moduleSessionLookupId;

            if (this.hasSession) {
                resolve();
            }

            const request = ProtocolGui.getRequestSqleditorStartSession();
            const listener = MessageScheduler.get.sendRequest(request,
                { messageClass: ShellAPIGui.GuiSqleditorStartSession });

            listener.then((event: ICommStartSessionEvent) => {
                if (event.data) {
                    webSession.setModuleSessionId(this.moduleSessionLookupId, event.data.result.moduleSessionId);

                    resolve();
                }
            }).catch((event: ICommErrorEvent) => {
                reject(event.message);
            });
        });
    }

    /**
     * Closes this editor session and all open connections.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public closeSession(): Promise<void> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                resolve();
            } else {
                const request = ProtocolGui.getRequestSqleditorCloseSession(id);
                MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorCloseSession })
                    .then(() => {
                        webSession.setModuleSessionId(this.moduleSessionLookupId);
                        resolve();
                    }).catch((event: ICommErrorEvent) => {
                        reject(event.message);
                    });
            }
        });
    }

    /**
     * Returns information of how the modules should be displayed in the gui.
     *
     * @returns A listener for the response
     */
    public getGuiModuleDisplayInfo(): ListenerEntry {
        const request = ProtocolGui.getRequestSqleditorGetGuiModuleDisplayInfo();

        return MessageScheduler.get.sendRequest(request, { messageClass: "getModuleDisplayInfo" });
    }

    /**
     * Returns true as this extension object holds the backend implementation of a gui module.
     *
     * @returns A listener for the response
     */
    public isGuiModuleBackend(): ListenerEntry {
        const request = ProtocolGui.getRequestSqleditorIsGuiModuleBackend();

        return MessageScheduler.get.sendRequest(request, { messageClass: "isGuiModuleBackend" });
    }

    /**
     * Opens the database connection.
     *
     * @param dbConnectionId The id of the db connection.
     *
     * @returns A listener for the response
     */
    public openConnection(dbConnectionId: number): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestSqleditorOpenConnection(dbConnectionId, id);

        return MessageScheduler.get.sendRequest(request, { messageClass: "openConnection" });
    }

    /**
     * Executes the given SQL.
     *
     * @param sql The sql command to execute.
     * @param params The module session the function should operate on.
     *
     * @returns A listener for the response
     */
    public execute(sql: string, params?: string[]): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            const response: IDispatchEvent<IGenericResponse> = {
                id: "",
                eventType: EventType.ErrorResponse,
                message: "No open session",
                context: { messageClass: "execute" },
                data: { requestState: { type: "ERROR", msg: "" } },
            };

            return ListenerEntry.resolve(response);
        }

        const request = ProtocolGui.getRequestSqleditorExecute(sql, id, params,
            { rowPacketSize: settings.get("sql.rowPacketSize", 1000) });

        return MessageScheduler.get.sendRequest(request, { messageClass: "execute" });
    }

    /**
     * Reconnects the database connection.
     *
     * @returns A listener for the response
     */
    public reconnect(): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestSqleditorReconnect(id);

        return MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorReconnect });
    }

    /**
     * Stops the currently running query (if there's any).
     *
     * @returns A listener for the response
     */
    public killQuery(): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestSqleditorKillQuery(id);

        return MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorKillQuery });
    }

    /**
     * Sets the auto commit mode for the current connection.
     * Note: this mode can implicitly be changed by executing certain SQL code (begin, set autocommit, rollback, etc.).
     *
     * @param value A flag indicating if the mode should be enabled or disabled.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public setAutoCommit(value: boolean): Promise<void> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve();
            }

            const request = ProtocolGui.getRequestSqleditorSetAutoCommit(id, value);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorSetAutoCommit })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve();
                    }
                })
                .catch((event) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Returns the current auto commit mode, if supported.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public getAutoCommit(): Promise<boolean> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve(false);
            }

            const request = ProtocolGui.getRequestSqleditorGetAutoCommit(id);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorGetAutoCommit })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve((event.data?.result as number) !== 0);
                    }
                })
                .catch((event) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Returns the current default schema, if supported.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public getCurrentSchema(): Promise<string> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve("");
            }

            const request = ProtocolGui.getRequestSqleditorGetCurrentSchema(id);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorGetCurrentSchema })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve(event.data ? event.data.result as string : "");
                    }
                }).catch((event: ICommErrorEvent) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Sets the current default schema, if supported.
     *
     * @param schema The schema to set as the default.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public setCurrentSchema(schema: string): Promise<void> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve();
            }

            const request = ProtocolGui.getRequestSqleditorSetCurrentSchema(id, schema);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorSetCurrentSchema })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve();
                    }
                }).catch((event: ICommErrorEvent) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Sends a reply from the user back to the backend (e.g. passwords, choices etc.).
     *
     * @param requestId The same request ID that was used to request input from the user.
     * @param type Indicates if the user accepted the request or cancelled it.
     * @param reply The reply from the user.
     *
     * @returns A listener for the response.
     */
    public sendReply(requestId: string, type: ShellPromptResponseType, reply: string): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestPromptReply(requestId, type, reply, id);

        return MessageScheduler.get.sendRequest(request, { messageClass: "sendReply" });
    }
}
