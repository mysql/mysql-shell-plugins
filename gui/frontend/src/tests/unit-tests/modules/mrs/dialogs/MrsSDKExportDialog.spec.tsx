/*
 * Copyright (c) 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

/* eslint-disable dot-notation */

import { createRef } from "preact";

import { mount } from "enzyme";
import { IMrsServiceData } from "../../../../../communication/ProtocolMrs.js";
import { MrsHub } from "../../../../../modules/mrs/MrsHub.js";
import { ShellInterfaceSqlEditor } from "../../../../../supplement/ShellInterface/ShellInterfaceSqlEditor.js";
import { MySQLShellLauncher } from "../../../../../utilities/MySQLShellLauncher.js";
import { KeyboardKeys } from "../../../../../utilities/helpers.js";
import {
    DialogHelper,
    JestReactWrapper,
    createBackend,
    recreateMrsData,
    sendKeyPress,
    setupShellForTests,
} from "../../../test-helpers.js";

describe("MRS SDK Export dialog tests", () => {
    let host: JestReactWrapper;
    let service: IMrsServiceData;
    let launcher: MySQLShellLauncher;
    const hubRef = createRef<MrsHub>();
    let dialogHelper: DialogHelper;
    let backend: ShellInterfaceSqlEditor;

    beforeAll(async () => {
        launcher = await setupShellForTests(false, true, "DEBUG2");

        const result = await recreateMrsData();
        service = result.service;

        host = mount<MrsHub>(<MrsHub ref={hubRef} />);

        dialogHelper = new DialogHelper("mrsSdkExportDialog", "Export MRS SDK for /myService");
    });

    afterAll(async () => {
        await backend.execute("DROP DATABASE IF EXISTS mysql_rest_service_metadata");
        await backend.execute("DROP DATABASE IF EXISTS MRS_TEST");
        await backend.closeSession();
        await launcher.exitProcess();
        host.unmount();
    });

    beforeEach(async () => {
        backend = await createBackend();
    });

    it("Show MRS SDK Export Dialog (snapshot) and escape", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
        const promise = hubRef.current!.showMrsSdkExportDialog(backend, service.id, 1);
        await dialogHelper.waitForDialog();


        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(1);

        expect(portals[0]).toMatchSnapshot();

        setTimeout(() => {
            sendKeyPress(KeyboardKeys.Escape);
        }, 250);

        await promise;

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });

    it("Show MRS SDK Export Dialog and cancel", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
        const promise = hubRef.current!.showMrsSdkExportDialog(backend, service.id, 1);
        await dialogHelper.waitForDialog();


        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(1);

        expect(portals[0]).toMatchSnapshot();

        await dialogHelper.clickCancel();

        await promise;

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });

    it("Dialog error testing", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
        const promise = hubRef.current!.showMrsSdkExportDialog(backend, service.id, 1);
        await dialogHelper.waitForDialog();


        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(1);

        await dialogHelper.clickOk();
        dialogHelper.verifyErrors();

        await promise;

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });
});
