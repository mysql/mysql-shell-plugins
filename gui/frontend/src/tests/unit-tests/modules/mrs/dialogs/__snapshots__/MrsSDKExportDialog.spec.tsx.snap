// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MRS SDK Export dialog tests Show MRS SDK Export Dialog (snapshot) and escape 1`] = `
<div
  class="msg portal"
  id="mrsSdkExportDialog"
  style="--background-opacity: 0.5;"
>
  <div
    class="msg dialog visible valueEditDialog"
  >
    <div
      class="title verticalCenterContent"
    >
      <div
        class="msg icon codicon codicon-terminal"
      />
      <label
        class="msg label"
        data-tooltip="expand"
      >
        Export MRS SDK for /myService
      </label>
      <div
        class="msg button imageOnly"
        id="closeButton"
        role="button"
        tabindex="0"
      >
        <div
          class="msg icon"
          style="mask-image: url("close2.svg");"
        />
      </div>
    </div>
    <div
      class="content fixedScrollbar"
    >
      <div
        class="msg grid"
        id="contentGrid"
        style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
      >
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: row; grid-column: span 8;"
        >
          <label
            class="msg label sectionTitle"
            data-tooltip="expand"
          >
            Export MRS SDK Files for /myService
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            REST Service URL
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="serviceUrl"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The URL to access the REST Service
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            SDK Client API Language
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar dropdown valueEditor"
              data-tooltip="inherit"
              id="sdkLanguage"
              style="flex-direction: row; align-items: center;"
              tabindex="0"
            >
              <label
                class="msg label ellipsis"
                data-tooltip="expand"
              >
                TypeScript
              </label>
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The development language that should be used for generation
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Include AppBaseClass
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar dropdown valueEditor"
              data-tooltip="inherit"
              id="addAppBaseClass"
              style="flex-direction: row; align-items: center;"
              tabindex="0"
            />
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              Add an application BaseClass with core MRS functionality.
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            SDK File Header
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <textarea
                class="msg input valueEditor"
                id="header"
                rows="4"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The header that should be applied to the generated SDK files.
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer verticalCenterContent"
    >
      <div
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <div
          caption="OK"
          class="msg button"
          id="ok"
          role="button"
          tabindex="0"
        >
          OK
        </div>
        <div
          caption="Cancel"
          class="msg button"
          id="cancel"
          role="button"
          tabindex="0"
        >
          Cancel
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MRS SDK Export dialog tests Show MRS SDK Export Dialog and cancel 1`] = `
<div
  class="msg portal"
  id="mrsSdkExportDialog"
  style="--background-opacity: 0.5;"
>
  <div
    class="msg dialog visible valueEditDialog"
  >
    <div
      class="title verticalCenterContent"
    >
      <div
        class="msg icon codicon codicon-terminal"
      />
      <label
        class="msg label"
        data-tooltip="expand"
      >
        Export MRS SDK for /myService
      </label>
      <div
        class="msg button imageOnly"
        id="closeButton"
        role="button"
        tabindex="0"
      >
        <div
          class="msg icon"
          style="mask-image: url("close2.svg");"
        />
      </div>
    </div>
    <div
      class="content fixedScrollbar"
    >
      <div
        class="msg grid"
        id="contentGrid"
        style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
      >
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: row; grid-column: span 8;"
        >
          <label
            class="msg label sectionTitle"
            data-tooltip="expand"
          >
            Export MRS SDK Files for /myService
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            REST Service URL
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="serviceUrl"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The URL to access the REST Service
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            SDK Client API Language
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar dropdown valueEditor"
              data-tooltip="inherit"
              id="sdkLanguage"
              style="flex-direction: row; align-items: center;"
              tabindex="0"
            >
              <label
                class="msg label ellipsis"
                data-tooltip="expand"
              >
                TypeScript
              </label>
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The development language that should be used for generation
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Include AppBaseClass
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar dropdown valueEditor"
              data-tooltip="inherit"
              id="addAppBaseClass"
              style="flex-direction: row; align-items: center;"
              tabindex="0"
            />
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              Add an application BaseClass with core MRS functionality.
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            SDK File Header
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <textarea
                class="msg input valueEditor"
                id="header"
                rows="4"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The header that should be applied to the generated SDK files.
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer verticalCenterContent"
    >
      <div
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <div
          caption="OK"
          class="msg button"
          id="ok"
          role="button"
          tabindex="0"
        >
          OK
        </div>
        <div
          caption="Cancel"
          class="msg button"
          id="cancel"
          role="button"
          tabindex="0"
        >
          Cancel
        </div>
      </div>
    </div>
  </div>
</div>
`;
